File: Algorithm Tests/Disjoint Set Test.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <bits/stdc++.h>
using namespace std;
int n, m, p[1000002];
vector<int> e;

int find(int x) {
    return x!= p[x]? p[x] = find(p[x]): p[x];
}
void unionn (int x, int y) {
    p[x] = y;
}
int main() {
    scanf("%d %d", &n, &m);
    for (int i = 1; i <= n; i++)
        p[i] = i;
    for (int i = 1, x, y; i <=m; i++) {
        scanf("%d %d", &x, &y);
        int fx = find(x), fy = find(y);
        if (fx != fy) {
            unionn(fx, fy);
            e.push_back(i);
        }
        if (e.size() == n - 1) break;
    }
    if (e.size() == n - 1)
        for (int u: e)
            printf("%d\n", u);
    else printf("Disconnected Graph");
}

File: Algorithm Tests/Disjoint Set Test.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static class FastScanner {
        private BufferedReader br;
        private StringTokenizer st;
        FastScanner(InputStream in) {
            br = new BufferedReader(new InputStreamReader(in));
            st = new StringTokenizer("");
        }

        private String next() throws IOException {
            if (st.hasMoreTokens()) return st.nextToken();
            else st = new StringTokenizer(br.readLine());
            return next();
        }

        public int nextInt() throws IOException {
            return Integer.parseInt(next());
        }

        public double nextDouble() throws IOException {
            return Double.parseDouble(next());
        }

        public long nextLong() throws IOException {
            return Long.parseLong(next());
        }

        public String nextLine() throws IOException {
            if (!st.hasMoreTokens()) return br.readLine();
            String cont = st.nextToken();
            while (st.hasMoreTokens()) {
                cont += " " + st.nextToken();
            }
            return cont;
        }
    }

    
    private static int p[];
    private static int find(int x) {
        if (x!= p[x])
            p[x] = find(p[x]);
        return p[x];
    }
    private static void union(int u, int v) {
        p[u] = v;
    }
    public static void main(String[] args) throws IOException {
        FastScanner sc = new FastScanner(System.in);
        int n = sc.nextInt(), m = sc.nextInt(), count = 0;
        ArrayList<Integer> arrL = new ArrayList<>();
        p = new int[n + 1];
        for (int i = 1; i <= n; i++)
            p[i] = i;
        for (int i = 1; i <= m; i++) {
            int a = sc.nextInt(), b = sc.nextInt();
            int fa = find(a), fb = find(b);
            if (fa != fb) {
                union(fa, fb);
                count++;
                arrL.add(i);
            }
            if (count == n-1) 
                break;
        }
        if (count == n-1) {
            for (int u: arrL)
                System.out.println(u);
        } else 
            System.out.println("Disconnected Graph");
        
        

    }
}